Add air-Q integration (air quality sensors)

Sibgatulin opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
A new component to integrate [air-Q](https://en.air-q.com/) devices with Home Assistant.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/23809

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!

  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.

  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

dl2080 committed:
Added initial files for air-Q integration

Sibgatulin committed:
Allow FIXME comments in pylint (temporary)

Also reintroduce --ignore-missing-annotations=y

Sibgatulin committed:
Set up air-q entry to connect to aioairq's API (initial attempt)

Also add necessary constants

Sibgatulin committed:
Implement a class for sensors and its update logic

Very early stage, WIP

Sibgatulin committed:
Zeroconf and authentication are working

Sibgatulin committed:
Complete the bare-bone minimal working version

Specifically, make AirQSensor update its values.

Sibgatulin committed:
Handle invalid authentication gracefully

Sibgatulin committed:
Handle ClientConnectionError gracefully

Sibgatulin committed:
Add field hint for the login form

The key in the schema, which defines the form in
`ConfigFlow.async_show_form` is looked up in both `airq/strings/json`
and `airq/translations/en.json`. I am still not 100% sure how this
lookup is performed. WIP

Sibgatulin committed:
Minor cleanups

Sibgatulin committed:
Extend sensor list to all supported by SensorDeviceClass

Also manage warming up sensors

Sibgatulin committed:
aioairq is published to PyPI and mentioned in requirements

Sibgatulin committed:
Reordered constants and list content alphabetically

As required by style guides. Also turned SENSOR_TYPES to a list

Sibgatulin committed:
Updated file docstrings for a dev unfamiliar w/homeassistant like myself

Sibgatulin committed:
Adding a bit of logging for the integration setup process

Sibgatulin committed:
Expose scan interval & smoothing flag

Also streamline test_authentication in config_flow.

Sibgatulin committed:
Fix a type annotation mistake

Sibgatulin committed:
Use as many constants from homeassistant.const as possible

My only concern is using CONST_IP_ADDRESS = "ip_address" for smth which
stands for both IP address and mDNS...

Sibgatulin committed:
Temporarily rollback ConfigFlow.async_step_configure and use defaults

TODO: implement OptionFlowHandler instead

Sibgatulin committed:
Define custom Coordinator, w subset of airq config

The latter is then accessed from entity / sensor constructors to define
correct DeviceInfo

Sibgatulin committed:
Provide translations to de & fr + minor changes to en

Sibgatulin committed:
Provide translations to ru + a minor en changes

Sibgatulin committed:
Make translation a little more helpful and polite

Sibgatulin committed:
Fix devicename and entry title

Sibgatulin committed:
Remove stale commented out code

Sibgatulin committed:
Test config_flow

At this point two helper functions which interact with the external
library are not tested

homeassistant commented:
Hi @Sibgatulin, @dl2080, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#76999).

Once you do that we will be able to review and accept this pull request.

Thanks!

Drafteed commented on .github/workflows/ci.yaml:
```diff
@@ -622,14 +622,14 @@ jobs:
           . venv/bin/activate
           python --version
-          pylint --ignore-missing-annotations=y homeassistant
+          pylint --ignore-missing-annotations=y homeassistant -d fixme
```
These changes are not related to integration and should not be located here.

Sibgatulin replied:
Thanks for spotting! It wasn't meant to be included. Fixed.

Drafteed commented on .github/workflows/ci.yaml:
```diff
@@ -622,14 +622,14 @@ jobs:
           . venv/bin/activate
           python --version
-          pylint --ignore-missing-annotations=y homeassistant/components/${{ needs.info.outputs.integrations_glob }}
+          pylint homeassistant/components/${{ needs.info.outputs.integrations_glob }}
```
Same as above.

Drafteed commented on .gitignore:
```diff
@@ -132,3 +132,7 @@ tmp_cache
 # python-language-server / Rope
 .ropeproject
+
+# own
```
Same as above.

Drafteed commented on .pre-commit-config.yaml:
```diff
@@ -96,7 +96,7 @@ repos:
       - id: pylint
         name: pylint
-        entry: script/run-in-env.sh pylint -j 0 --ignore-missing-annotations=y
+        entry: script/run-in-env.sh pylint -j 0 --ignore-missing-annotations=y -d fixme
```
Same as above.

emontnemery commented on homeassistant/components/airq/__init__.py:
```diff
@@ -0,0 +1,99 @@
+        self.airq = AirQ(address, passw)
+        self._target_route = target_route
+
+    # Consider adding a more specific type alias, e.g.
+    # Data = int | float | list[float] | str | dict[str, str]
+    # or filtering the dictionary in a dedicated method e.g. aioairq.AirQ.get_lean_data
```
I don't think we want this kind of general TODO-type comment in the code.

Sibgatulin replied:
I went over the PR and cleaned up such comments, which were intended for our internal use and which slipped through. Sorry about it.

emontnemery commented on homeassistant/components/airq/__init__.py:
```diff
@@ -0,0 +1,99 @@
+        passw=entry.data[CONF_PASSWORD],
+    )
+
+    # A workaround. Should be set in AirQCoordinator.__init__, I just failed
+    # to handle the async / await correctly
+    config = await coordinator.airq.get("config")
+    coordinator.config = {
+        "id": config["id"],
+        "name": config["devicename"],
+        "model": config["type"],
+        "room_type": config["RoomType"].replace("-", " ").title(),
+        "sw_version": config["air-Q-Software-Version"],
+        "hw_version": config["air-Q-Hardware-Version"],
+    }
```
You can't `await` in `__init__`. What you *can* do however, is to add an async method `async def async_setup` to 
`AirWCoordinator` and call that after you've created the object.

Sibgatulin replied:
Thanks for the suggestion! Done.

emontnemery commented on homeassistant/components/airq/__init__.py:
```diff
@@ -0,0 +1,99 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up air-Q from a config entry."""
+    # entry.data is a dict with the data from STEP_USER_SCHEMA:
+    # entry.data.keys: [CONF_IP_ADDRESS, CONF_PASSWORD]
```
Instead of this comment, you can define a typed dict `AirQData` and do:
```py
airq_data: AirQData = entry.data
```

Sibgatulin replied:
Thanks for the idea. While I like it in general, I stopped seeing the need for such clarification anymore and thus removed the comment.

emontnemery commented on homeassistant/components/airq/config_flow.py:
```diff
@@ -0,0 +1,123 @@
+        try:
+            await validate_input(user_input)
+        except CannotConnect:
+            # Is it ok to print the IP? Looks so, judging by this
+            # https://developers.home-assistant.io/docs/development_guidelines?_highlight=print&_highlight=out#log-messages
+            # Also, I am opting for .error over .exception since the latter includes
+            # a humongous traceback, which looks far too scary for such a minor thing
```
Instead of asking questions in comments, please to your own review of the code where you ask questions or add some more detailed explanations which shouldn't be in comments.

emontnemery commented on homeassistant/components/airq/config_flow.py:
```diff
@@ -0,0 +1,123 @@
+            # https://developers.home-assistant.io/docs/development_guidelines?_highlight=print&_highlight=out#log-messages
+            # Also, I am opting for .error over .exception since the latter includes
+            # a humongous traceback, which looks far too scary for such a minor thing
+            _LOGGER.error(
```
This should be `debug` or `info` since the user will get feedback in the config flow already 👍

Sibgatulin replied:
Fixed

emontnemery commented on homeassistant/components/airq/config_flow.py:
```diff
@@ -0,0 +1,123 @@
+                "Incorrect password for device %s", user_input[CONF_IP_ADDRESS]
+            )
+            errors["base"] = "invalid_auth"
+        except Exception:  # pylint: disable=broad-except
+            # This really shouldn't happen, so .exception is perhaps more appropriate
+            _LOGGER.exception("Unexpected exception")
+            errors["base"] = "unknown"
```
You shouldn't have to do this, the base config flow handler should handle it and give a reasonable error message to the user already.
Try raising some arbitrary exception and see what happens.

Sibgatulin replied:
Perhaps, I am missing something here. Simply omitting this except-clause fails `tests.components.airq.test_config_flow.test_form_unknown_error`. Am I testing a wrong thing there?

emontnemery commented on homeassistant/components/airq/manifest.json:
```diff
@@ -0,0 +1,19 @@
+  "documentation": "https://www.home-assistant.io/integrations/airq",
+  "requirements": ["aioairq==0.1.1"],
+  "ssdp": [],
+  "zeroconf": [
+    {
+      "type": "_http._tcp.local.",
+      "name": "*_air-q.local"
+    }
+  ],
```
There's no zeroconf step in the config flow?

Sibgatulin replied:
Thanks for spotting it. Unfortunately, I had to retract this as our zeroconf support is not yet quite ready.

emontnemery commented on homeassistant/components/airq/translations/de.json:
```diff
@@ -0,0 +1,23 @@
+{
+    "config": {
+        "abort": {
+            "already_configured": "Dieses Gerät ist bereits vorhanden und steht Home Assistant zur Verfügung"
+        },
+        "error": {
+            "cannot_connect": "Verbindung fehlgeschlagen, bitte IP oder mDNS überprüfen",
+            "invalid_auth": "Passwort falsch",
+            "unknown": "Unvorhergesehener Fehler"
+        },
+        "step": {
+            "user": {
+                "data": {
+                    "ip_address": "Geräte-IP oder mDNS (z.B. '123ab_air-q.local')",
+                    "password": "Passwort"
+                },
+                "description": "Netzwerk-Adresse (IP oder mDNS) des Gerätes und Geräte-Passwort",
+                "title": "Identifikation des Gerätes"
+            }
+        }
+    }
+}
+
```
Only en.json should be included in the PR, other files will be automatically generated.

emontnemery reviewed (changes requested):
Looks pretty good overall.

A few initial comments to take care of though.

emontnemery commented:
@Sibgatulin, is @dl2080 aware of being added as code owner?

dl2080 commented:
> @Sibgatulin, is @dl2080 aware of being added as code owner?

I am aware of being added as code owner :)

Sibgatulin committed:
Clean up unrelated and meant as temporary changes

Sibgatulin committed:
Clean up unnecessary comments meant for internal use

Sibgatulin committed:
Move fetching config to a dedicated async coordinator method

As opposed to it being a potentially poorly justified step in
async_setup_entry

Sibgatulin committed:
Remove zeroconf support since it is not yet ready

Sibgatulin committed:
Remove translations other than en

Drafteed commented on homeassistant/components/airq/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/airq",
+  "requirements": ["aioairq==0.1.1"],
+  "ssdp": [],
+  "homekit": {},
+  "dependencies": [],
```
Remove these empty sections as well

Drafteed commented on homeassistant/components/airq/sensor.py:
```diff
@@ -0,0 +1,205 @@
+
+        # .get(key, None) over [key] to guard against the keys of the warming up
+        # sensors not being present in the returned dictionary
+        return self.coordinator.data.get(self.entity_description.key, None)
```
`get()` returns default value `None` if key is not available.

Sibgatulin replied:
Thanks for the hint!

Drafteed commented on homeassistant/components/airq/sensor.py:
```diff
@@ -0,0 +1,205 @@
+    @property
+    def native_value(self) -> float | int | None:
+        """Return the value reported by the sensor."""
+        # airthings has a neat way of doing it when the data returned by the API
+        # are a dictionary with keys for each device, and values being dictionaries
+        # of sensor values. Under this condition, the call should be:
+        # return self.coordinator.data[self._id].sensors[self.entity_description.key]
+        # In our case now, only one device is allowed and self.coordinator.data
+        # contains the regular dict retrieved from a single device
+
+        # .get(key, None) over [key] to guard against the keys of the warming up
+        # sensors not being present in the returned dictionary
```
I think this comment are redundant. Or you can make it more shorter.

Drafteed commented on homeassistant/components/airq/config_flow.py:
```diff
@@ -0,0 +1,119 @@
+        auth_success = await AirQ(
+            data[CONF_IP_ADDRESS], data[CONF_PASSWORD]
+        ).test_authentication()
+    except ClientConnectionError as exc:
+        raise CannotConnect from exc
```
Catch `ClientConnectionError` at top level instead inheritance.

Sibgatulin replied:
Right, thanks! Done.

Sibgatulin committed:
Remove unnecessary comments, manifest.json entries, and constants

Sibgatulin committed:
Improve exception handling

- `_LOGGER` uses `debug` and not `error` levels.
- Drop `ClientConnect` and catch `aiohttop.ClientConnectError` directly
- Drop `Exception` as it is not expected from `aioairq` (remove the
  corresponding test too)

Sibgatulin commented on homeassistant/components/airq/config_flow.py:
```diff
@@ -0,0 +1,119 @@
+            _LOGGER.error(
+                "Incorrect password for device %s", user_input[CONF_IP_ADDRESS]
+            )
+            errors["base"] = "invalid_auth"
```
Sorry for missing this one in prev commits. I am assuming, `_LOGGER.debug` is appropriate still?

Drafteed replied:
`_LOGGER.debug` is ok.

Sibgatulin commented:
Thanks for prompt and helpful comments @Drafteed and @emontnemery!

Drafteed commented on homeassistant/components/airq/__init__.py:
```diff
@@ -0,0 +1,96 @@
+        Additionally, the result dictionary is stripped of the errors. Subject to
+        a discussion
+        """
+        data = await self.airq.get(self._target_route)
+        return self.airq.drop_errors_from_data(data)
```
Some words about upstream library:
1. I noticed that you are creating a session every time you request. It would be better to create the session once (or get existed) using HA helper and pass into your library. ([example](https://github.com/home-assistant/core/blob/f3e432c9c7735daddd2946773cc6572567821a5e/homeassistant/components/adguard/__init__.py#L48-L57), [example](https://github.com/home-assistant/core/blob/f3e432c9c7735daddd2946773cc6572567821a5e/homeassistant/components/accuweather/__init__.py#L38-L42), [example](https://github.com/home-assistant/core/blob/f3e432c9c7735daddd2946773cc6572567821a5e/homeassistant/components/braviatv/__init__.py#L27-L30))
2. Instead of dropping errors (drop_errors_from_data), it is customary to throw an UpdateFailed exception here ([example](https://github.com/home-assistant/core/blob/f3e432c9c7735daddd2946773cc6572567821a5e/homeassistant/components/braviatv/coordinator.py#L124-L127)). However, aiohttp and some commons exceptions can be caught automatically by the DataUpdateCoordinator if they are not caught at the library level. See [DataUpdateCoordinator](https://github.com/home-assistant/core/blob/f3e432c9c7735daddd2946773cc6572567821a5e/homeassistant/helpers/update_coordinator.py#L214-L219).

Sibgatulin replied:
1. Thanks for the suggestion. It sounds good and I will look into it.
2. Perhaps, `drop_errors_from_data` is a misnomer, as by errors this method means uncertainties and not exceptions. Specifically, it drops uncertainties of the measurement values, which are estimated by the device and returned by `self.airq.get()`. I'll rename the method.

Sibgatulin replied:
Hey @Drafteed, I have refactored the upstream module to accept a session, and made according changes. Thank you for the suggestion!
I also renamed `drop_errors_from_data` method and added a formerly missing docstring.

Drafteed commented on homeassistant/components/airq/__init__.py:
```diff
@@ -0,0 +1,96 @@
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```
```suggestion
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
```

Drafteed commented on homeassistant/components/airq/strings.json:
```diff
@@ -0,0 +1,30 @@
+          "password": "[%key:common::config_flow::data::password%]"
+        }
+      },
+      "configure": {
+        "title": "Configure the integration",
+        "description": "Select the desired scan interval (in seconds) and if the data should be smoothed",
+        "data": {
+          "scan_interval": "Scan interval in seconds",
+          "show_average": "Smooth the momentary values"
+        }
+      }
```
This step is missing.

Drafteed commented:
Please resolve outdated comments to improve readability.

Sibgatulin committed:
Drop strings for obsolete errors and steps

Specifically, `unknown` error isn't caught any more. `configure` step
has also been removed.

Sibgatulin committed:
Refactor en.json to be consistent with strings.json

Sibgatulin committed:
Move target_route from a coordinator argument to a constant

At this point a user cannot configure the target_route route, thus it
does not make sense to expose it half-heartedly in
`AirQCoordinator.__init__`, since it cannot be accessed.

Sibgatulin committed:
Fix an async call in `AirQCoordinator.async_setup_entry`

Sibgatulin commented on homeassistant/components/airq/__init__.py:
```diff
@@ -0,0 +1,96 @@
+        update_interval: timedelta,
+        address: str,
+        passw: str,
+        target_route: str = "average",
```
Was left as a loose end for future enhancement. Indeed, unnecessary now. Used a constant instead.

Sibgatulin committed:
Refactor underlying aioairq API

- Use `homeassistant.helpers.aiohttp.async_get_clientsession` and pass a
  single persistent session to `aioariq.AirQ.__init__`
- `aioairq.AirQ.fetch_device_info` now returns a `DeviceInfo` object heavily
  inspired and almost compatible with `homeassistant.helpers.entity.DeviceInfo`.
  Make heavier use of this object and define a single `DeviceInfo` in the
  `AirQCoordinator` (instead of recreating the same object for each sensor
  of the device in `sensor.AirQSensor`)
- Drop two helper functions in `config_flow.py` and operate on `aioariq.AirQ`
  methods directly

Sibgatulin commented:
I shouldn't have rebased just yet, should I?.. :grimacing:

l-mb commented:
As a potential user of this I thought I'd give it a try to review and it looks good to me. (Except for the complaints in the CODEOWNERS file that I don't know how to resolve.)

I tried the `aioairq` library and it works fine, but it also captures significantly more metrics than are included in `SENSOR_TYPES`: sound, sound_max, oxygen, dewpt, and also health and performance metrics. If you're doing one more round, I'd recommend to either list those that are currently dropped in the documentation, or to - preferably, from the user's point of view :-) - to add them in.

Sibgatulin committed:
Fix the version of aioairq

Sibgatulin commented:
Hi @l-mb! Thanks for taking care and having a careful look!

> (Except for the complaints in the CODEOWNERS file that I don't know how to resolve.)

I must say that I am too rather perplexed about these complaints. Maybe @emontnemery or @Drafteed could shed some light?

You are spot on regarding the presence of additional sensor readings and metrics. In this initial PR we thought to limit ourselves to the sensors types already present in `homeassistant.components.sensor.SensorDeviceClass`. 
I could not find any notes or guidelines on adding members to this enumeration. If acceptable, I assume oxygen and sound, for example, would be sufficiently general types of sensors to be included. 
Most of the other sensor types provided by the device may be somewhat a little more niche. However, now that I think of it and compare with other integrations, I see a practice of defining a `SensorEntityDescription` with a custom `device_class` (or simply `None`), when an appropriate type is missing. I think we can add more entities following this path. Thanks for the suggestion!

Sibgatulin committed:
Add 15 more sensors + icons

Sibgatulin committed:
Remove cnt* & TypPS, change units of health & performance

Sibgatulin committed:
Add 12 more sensors

Sibgatulin committed:
Add a missing icon

Sibgatulin committed:
Gracefully handle device not being available on setup

If the device and the host are not on the same WiFi,
ServerTimeoutError is raised, which is caught by
ClientConnectionError.
If the device is powered off, ClientConnectionError is expected.
In both cases, ConfigEntryNotReady is raised, as prescribed by the docs.
Newer version of aioairq times-out far quicker than the default 5 mins.

Sibgatulin committed:
Rename two sensors

Sibgatulin committed:
Validate provided IP address / mDNS

aioairq now raises InvalidInput if provided IP / mDNS does not seem
valid. Handle this exception correctly

Sibgatulin committed:
Merge branch 'extra_sensors' into air_q

Sibgatulin committed:
Merge branch 'validate_input' into air_q

emontnemery commented:
>> (Except for the complaints in the CODEOWNERS file that I don't know how to resolve.)

>I must say that I am too rather perplexed about these complaints. Maybe @emontnemery or @Drafteed could shed some light?

If you mean the read squigglies, don't worry about that; they just mean that github users without access to `home-assistant/core` are added as code owners.

Sibgatulin commented:
> If you mean the read squigglies, don't worry about that; they just mean that github users without access to `home-assistant/core` are added as code owners.

Alright, thanks for the clarification.

emontnemery reviewed (changes requested):
This looks pretty good now, just drop the custom device classes 👍

Sibgatulin committed:
Apply suggestions from code review

Clean up the comments and rename the logger

Co-authored-by: Erik Montnemery <erik@montnemery.com>
Co-authored-by: Artem Draft <Drafteed@users.noreply.github.com>

Drafteed commented on homeassistant/components/airq/__init__.py:
```diff
@@ -0,0 +1,102 @@
+        passw=entry.data[CONF_PASSWORD],
+    )
+
+    try:
+        await coordinator.async_fetch_device_info()
+    except ClientConnectionError as ex:
+        raise ConfigEntryNotReady(
+            f"Failed to connect to device {entry.data[CONF_IP_ADDRESS]}. "
+            "Check if the device is on and connected to the same WiFi as the host"
+        ) from ex
```
I don’t seem to understand very well, why do we need to request from API information about the device every time we start?
It seems to me that you can save the necessary data at the config flow step. I don't think it should be here at all.
Besides, it is not enough to catch only `ClientConnectionError`. There are other exceptions that can be thrown during a request, such as Timeout or Connection Reset.

emontnemery commented:
I should clarify a bit about custom device classes:

It's perfectly fine to add custom device classes, as long as they're prefixed with `airq__`.
That also allows you to add translations in file `homeassistant/components/airq/strings.sensor.json`.
Sorry about my misleading comment earlier.

Sibgatulin committed:
Only fetch device info during the first refresh

- Fetched info is stored in AirQCoordinator.device_info.
- In `AirQSensor.native_value` only multiply by the factor if
  the sensor reading is not None
- Fix the tests for ConfigFlow for aioairq==0.2.3. Specifically
  make the dummy data pass the new validation step upstream
  + add a test which fails it

Sibgatulin committed:
Drop custom device classes for now

Sibgatulin commented on homeassistant/components/airq/__init__.py:
```diff
@@ -0,0 +1,102 @@
+
+        Function is meant as an async closure, or partial(airq.get, TARGET_ROUTE)
+        Additionally, the result dictionary is stripped of the errors. Subject to
+        a discussion
```
Removed

Drafteed commented on homeassistant/components/airq/sensor.py:
```diff
@@ -0,0 +1,324 @@
+        description for description in SENSOR_TYPES if description.key in available_keys
+    ]
+    _LOGGER.debug(
+        "Identified %d  available sensors: %s",
```
```suggestion
        "Identified %d available sensors: %s",
```

Sibgatulin committed:
Apply suggestions from code review

Co-authored-by: Artem Draft <Drafteed@users.noreply.github.com>

Sibgatulin committed:
Only fetch device info during ConfigFlow.async_step_user

Store the result obtained by `airq.fetch_device_info` it in
`config_entry.data`. Pass the entire config entry to `AirQCoordinator`
and build the entire `homeassistant.helpers.entity.DeviceInfo` in the
`AirQCoordinator.__init__`. This makes
`AirQCoordinator._async_fetch_device_info` and overloaded
`AirQCoordinator._async_config_entry_first_refresh` obsolete.

Bump aioairq version.

Turn update_interval from `AirQCoordinator.__init__` argument into a
contestant.

Sibgatulin committed:
Merge branch 'air_q' of github.com:CorantGmbH/home-assistant-core into air_q

Sibgatulin commented on homeassistant/components/airq/__init__.py:
```diff
@@ -0,0 +1,107 @@
+        self.device_id = device_info.pop("id")
+
+        self.device_info["suggested_area"] = device_info.pop("room_type")
+        self.device_info["identifiers"] = {(DOMAIN, self.device_id)}
```
I somehow fail to pacify mypy, which appears most concerned with `entity.unique_id` potentially being a None. I must be missing something, but currently, as much as I don't like it, I feel I still need to duplicated `device_id`/`unique_id`. Would you have any suggestions?

Drafteed replied:
> which appears most concerned with `entity.unique_id` potentially being a None

```py
assert entity.unique_id
not_none = entity.unique_id
```

Drafteed replied:
Example https://github.com/home-assistant/core/blob/0ce526efe1ab780b0dae88430378ee206f178f22/homeassistant/components/braviatv/media_player.py#L26-L32

Sibgatulin replied:
Right! Thank you, done

Sibgatulin commented on homeassistant/components/airq/__init__.py:
```diff
@@ -0,0 +1,107 @@
+        update_interval=timedelta(seconds=10),
+        address=entry.data[CONF_IP_ADDRESS],
+        passw=entry.data[CONF_PASSWORD],
+    )
```
Done, thanks for the tips

Drafteed commented on homeassistant/components/airq/sensor.py:
```diff
@@ -0,0 +1,324 @@
+        self._attr_name = description.name
+        self._attr_unique_id = f"{coordinator.device_id}_{description.key}"
+        # the following two sensor units must be converted to %
+        self._factor = 0.1 if description.key in ["performance", "health"] else 1.0
```
You can create your own dataclass description based on `SensorEntityDescription` which will contain a lambda function that returns the value of the sensor. In this function, you can transform your value, for example divide it. This is good practice here.
https://github.com/home-assistant/core/blob/ac3534cba120688b1d40d1b6bbc947322aa8c776/homeassistant/components/adguard/sensor.py#L25-L53

Sibgatulin replied:
Sounds good, thanks for the suggestion. 
I chose to expose a relatively lean lambda allowing to transform already retrieved value (while in `adguard` `value_fn` appears to also take on the job of retrieving the data from `AdGuardHome`). Hope it's ok.

Sibgatulin replied:
In such approach I am a little concerned with the need to duplicate the key of each sensor. Exposing the key in the lambda could be an option, which would land us here:

```python
@dataclass
class AirQEntityDescription(SensorEntityDescription):
    """Describes AirQ sensor entity. """

    value_fn: Callable[
        [AirQCoordinator, str], float | int | None
    ] = lambda coordinator, key: coordinator.data.get(key)

SENSOR_TYPES: list[AirQEntityDescription] = [
    AirQEntityDescription(
        key="nh3_MR100",
        name="Ammonia",
        native_unit_of_measurement=CONCENTRATION_MICROGRAMS_PER_CUBIC_METER,
        state_class=SensorStateClass.MEASUREMENT,
    ),
    AirQEntityDescription(
        key="health",
        name="Health Index",
        native_unit_of_measurement=PERCENTAGE,
        state_class=SensorStateClass.MEASUREMENT,
        icon="mdi:heart-pulse",
        value_fn=lambda coordinator, key: coordinator.data.get(key) / 10.0
    ),
]

@property
def native_value(self) -> float | int | None:
    """Return the value reported by the sensor."""
    return self.entity_description.value_fn(
        self.coordinator, self.entity_description.key
    )
```

However, I think I do not immediately see enough reasons to factor `coordinator.data.get(key)` out out of `native_value` into a lambda, since at present it is the step shared by all sensors. I would thus refrain from introducing `value_fn` as a function which retrieves desired value in the most general case given the coordinator, but instead view it as an optional transform which can be applied to selected (few) values. Perhaps, it can be renamed to `value_transform_fn` or smth.

Drafteed replied:
As an alternative:
```py
@property
def native_value(self) -> float | int | None:
    """Return the value reported by the sensor."""
    if self.entity_description.value_fn:
        return self.entity_description.value_fn(self.coordinator)
    return self.coordinator.get(self.entity_description.key)
```

But I think renaming to `value_transform_fn`  is also ok.

Drafteed replied:
Still, it's probably better to get data in a lambda function. 
Check out the implementation https://github.com/home-assistant/core/blob/dev/homeassistant/components/nextdns/sensor.py

You also need to check if your sensor data is updated in real time, as you may need set callback the coordinator as in the example above.

Sibgatulin replied:
Thanks for the example. 

> You also need to check if your sensor data is updated in real time

I am not 100% sure where and how to best check (i.e. assert in the code?). Yet I swapped the `native_value` property for a `callback` as in the example. I hope that's correct.

> Still, it's probably better to get data in a lambda function.

Latest commits make it so, but I am afraid that in this case I still find it hard to justify the duplication of the keys and repetition of the same call in every entity description instance (also in the example of nextdns). I probably miss something, in which case please let me know :)

Drafteed replied:
Look good now!

Drafteed commented on homeassistant/components/airq/config_flow.py:
```diff
@@ -0,0 +1,89 @@
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial configuration step: authentication.
+
+        This prompts the user to enter the credentials to access the device.
+        It then tries to connect to the device, and to assess the validity
+        of the password.
```
```suggestion
        """Handle the initial (authentication) configuration step."""
```

Drafteed commented on homeassistant/components/airq/sensor.py:
```diff
@@ -0,0 +1,324 @@
+    def native_value(self) -> float | int | None:
+        """Return the value reported by the sensor."""
+        # While a sensor is warming up its key isn't present in the returned dict
+        value = self.coordinator.data.get(self.entity_description.key)
+        return value * self._factor if value else value
```
Related to comment above.

Drafteed reviewed (commented):
Now it looks good 👍. My final comments.

Sibgatulin committed:
Custom entity description exposing a hook to modify sensor value

Use a `AirQEntityDescription` with a callable `value_fn` which allows to
change the sensor value retrieved from the device. Note that the
callable does not handle data retrieval itself (even from
`coordinator.data`). Instead it is purely a hook to transform obtained
value.

Sibgatulin committed:
Avoid duplicated use of unique_id

Device info is fetched  during the `ConfigFlow.async_user_step`.
`unique_id` is taken from the device info and is **not** stored
in `config_entry.data`. Subsequently `config_entry.unique_id` is
used instead.

Drafteed commented on homeassistant/components/airq/__init__.py:
```diff
@@ -0,0 +1,82 @@
+
+    async def _async_update_data(self) -> dict:
+        """Fetch the data from the device."""
+        try:
+            data = await self.airq.get(TARGET_ROUTE)
+        except ClientError as err:
+            raise UpdateFailed(f"Error while retrieving data: {err}") from err
+        return self.airq.drop_uncertainties_from_data(data)
```
Catching `ClientError` is redundant here. As I said earlier, some common exceptions, including `aiohttp` errors, will be [caught by the coordinator automatically](https://github.com/home-assistant/core/blob/f3e432c9c7735daddd2946773cc6572567821a5e/homeassistant/helpers/update_coordinator.py#L204-L219).
```suggestion
        data = await self.airq.get(TARGET_ROUTE)
        return self.airq.drop_uncertainties_from_data(data)
```

Sibgatulin replied:
Right, thanks!

Sibgatulin committed:
Drop unnecessary try-except

Co-authored-by: Artem Draft <Drafteed@users.noreply.github.com>

Sibgatulin committed:
Clarify the use of value_transform_fn

Sibgatulin committed:
Refactor the use of lambdas in AirQEntityDescription

Now it is the job of the callable under `value` to get the sensor
reading from the coordinator's data. Factoring this functionality into a
callback decouples  the key of the description from the key of dict,
returned by the API, so `AirQEntityDescription` no longer requires its key
to be set to smth clearly internal (e.g. `nh3_MR100`).

Sibgatulin committed:
Use a callback to update native_value

Since all `native_value`s are updated synchronously, it can as well be
done in a callback for better state consistency (right?)

Sibgatulin committed:
Revert the description keys to match data keys

Must match given the current way of identifying available sensors. On a
broader scale, they must match to be able to relate the descriptions to
sensors, unless a separate lookup table is maintained.

Drafteed commented on homeassistant/components/airq/sensor.py:
```diff
@@ -0,0 +1,369 @@
+    available_keys = list(coordinator.data.keys())
+
+    # Add sensors under warmup
+    status = coordinator.data["Status"]
+    if isinstance(status, dict):
+        warming_up_sensors = [
+            k for k, v in status.items() if "sensor still in warm up phase" in v
+        ]
+        available_keys.extend(warming_up_sensors)
+        _LOGGER.debug(
+            "Following %d sensors are warming up: %s",
+            len(warming_up_sensors),
+            ", ".join(warming_up_sensors),
+        )
```
Can you clarify what it means `sensor still in warm up phase` and what this section of code does. Did I understand correctly that the keys of some sensors may be missing in coordinator data if they are not ready and this section of code adds them?

Sibgatulin replied:
Exactly this. After a device restart, some sensors take time to get ready and start returning values. Until then the respective keys are missing in `coordinator.data` but instead present in `coordinator.data["Status"]` (which looks approximately like `{sensor_name: f"{sensor_name} sensor still in warm up phase; waiting time = %d s"}`).

Drafteed replied:
Let's reduce the number of iterations, eg:

```py
async def async_setup_entry(
    hass: HomeAssistant,
    config: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up sensor entities based on a config entry."""

    coordinator = hass.data[DOMAIN][config.entry_id]

    entities: list[AirQSensor] = []
    available_sensors: list[str] = []
    warming_sensors: list[str] = []

    for description in SENSOR_TYPES:
        if description.key not in coordinator.data:
            status = coordinator.data["Status"].get(description.key)
            if not status or "sensor still in warm up phase" not in status:
                continue
            warming_sensors.append(description.key)
        available_sensors.append(description.key)
        entities.append(AirQSensor(coordinator, description))

    _LOGGER.debug(
        "Following %d sensors are warming up: %s",
        len(warming_sensors),
        ", ".join(warming_sensors),
    )
    _LOGGER.debug(
        "Identified %d sensors: %s",
        len(available_sensors),
        ", ".join(available_sensors),
    )

    async_add_entities(entities)
```

Drafteed replied:
Or even better:

```py
async def async_setup_entry(
    hass: HomeAssistant,
    config: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up sensor entities based on a config entry."""

    coordinator = hass.data[DOMAIN][config.entry_id]

    entities: list[AirQSensor] = []

    for description in SENSOR_TYPES:
        if description.key not in coordinator.data:
            status = coordinator.data["Status"].get(description.key)
            if not status or "sensor still in warm up phase" not in status:
                continue
            _LOGGER.debug("Following sensor is warming up: %s", description.key)
        entities.append(AirQSensor(coordinator, description))

    async_add_entities(entities)
```

Sibgatulin committed:
Reduce number of loops when adding sensors

Filtering warming up sensors and non-sensor keys can be combined with
adding entities.

Drafteed reviewed (approved):
Looks good to me 👍 Maybe @emontnemery has something to say.

Sibgatulin commented:
Many thanks for all the input, @Drafteed!

l-mb commented:
Thanks again for all the work. I spent the weekend looking into this and it seems to be working fine!

Hopefully it'll show up in my next poll soon :)

Sibgatulin commented:
Thanks for looking into it carefully @l-mb!

emontnemery reviewed (approved):
LGTM, thanks @Sibgatulin 👍

dl2080 committed:
Merge branch 'dev' into air_q

Sibgatulin committed:
Remove obsolete imports

Sibgatulin commented:
Regarding the missing documentation label, I added the missing link to the original comment. This PR is also cross-referenced in the [corresponding](https://github.com/home-assistant/home-assistant.io/pull/23809) documentation PR. Hope it's alright now.

dl2080 committed:
Merge branch 'dev' into air_q

l-mb commented:
I don't want to be pushy, but was wondering if there was anything missing here that I could help with? I'd love to see this merged :)

Sibgatulin commented:
I was just wondering, if it is merely a CI workflow which a a maintainer needs to approve for running, or am I missing anything which must be done on the contributor side?

ElmoWater committed:
Merge branch 'dev' into air_q

Sibgatulin committed:
Merge branch 'home-assistant:dev' into air_q

Sibgatulin committed:
Update integrations.json

Sibgatulin committed:
Merge branch 'home-assistant:dev' into air_q

Sibgatulin committed:
Merge branch 'home-assistant:dev' into air_q

Sibgatulin committed:
Add integration_type

Integration supports multiple devices => hub

farmio reviewed (approved):
LGTM! thanks 👍

MartinHjelmare commented on homeassistant/components/airq/__init__.py:
```diff
@@ -0,0 +1,78 @@
+
+    async def _async_update_data(self) -> dict:
+        """Fetch the data from the device."""
+        data = await self.airq.get(TARGET_ROUTE)
```
Future improvement: We probably want to handle at least invalid authentication here and raise `ConfigEntryAuthFailed` and add a reauth step in the config flow.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#reauthentication

The parent coordinator class will handle aiohttp errors by default but if there are other library specific errors raised here we should catch them and raise `UpdateFailed` if it's not bad credentials.

MartinHjelmare commented on homeassistant/components/airq/config_flow.py:
```diff
@@ -0,0 +1,84 @@
+                )
+
+                device_info = await airq.fetch_device_info()
+                await self.async_set_unique_id(device_info.pop("id"))
```
What kind of ID is `id` here? Is it unique and stable?

Sibgatulin replied:
It is indeed a unique and constant 32-char serial number of the device, being configured.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!

Sibgatulin commented:
Thanks @MartinHjelmare for the detailed feedback and the comments! I'll try and address them all.
